hs/src/Reach/Optimize.hs

```diff
@@ -920,11 +926,72 @@ preFloat = \case
     void $ updateVarSet dv $ Just . maybe 1 succ
   _ -> return ()
 
--- Optimize is generally top-down, but this is locally bottom-up. We float after optimizing the tail
+updateFusion :: DLVar -> SwitchCases DLTail -> App ()
+updateFusion dv scs = do
+  fusionCasesR <- asks eFusionCases
+  liftIO $ modifyIORef' fusionCasesR $ M.insert dv scs
+
+readFusionCases :: DLVar -> App (Maybe (SwitchCases DLTail))
+readFusionCases dv = do
+  fusionCases <- liftIO . readIORef =<< asks eFusionCases
+  return $ M.lookup dv fusionCases
+
+preFuse :: DLStmt -> App (M.Map DLVar (SwitchCases DLTail))
+preFuse s = do
+  oldEnv <- liftIO . readIORef =<< asks eFusionCases
+  case s of
+    DL_LocalSwitch _ dv _ -> updateFusion dv mempty
+    DL_Let {} -> cantFuse
+    DL_Var {} -> cantFuse
+    _ | not (isPure s) -> cantFuse
+    _ -> return ()
+  return oldEnv
+  where cantFuse = liftIO . flip writeIORef mempty =<< asks eFusionCases
+
+fuseSwitch :: (DLStmt -> a -> a) -> SrcLoc -> DLVar -> SwitchCases DLTail -> a -> App a
+fuseSwitch mk at dv scs k =
+  readFusionCases dv >>= \case
+    Just cases -> ret $ M.mapWithKey (fuse cases) scs
+    Nothing -> ret scs
+  where
+    ret cases = return $ mk (DL_LocalSwitch at dv cases) k
+    fuse cases c (v1, b1, t1) =
+      case M.lookup c cases of
+        Just (v2, _, t2) -> do
+          -- The branch we're inling binds `v2`.
+          -- The branch we're inling *into* binds `v1`.
+          -- Substitute occurrences of `v2` in `t2` with `v1`.
+          let t2' = subst_ (M.fromList [(v2, v1)]) t2
+          (v1, b1, dtReplace DT_Com t2' t1)
```

Comment:
I think it is simpler to introduce `v2 := v1` and let other optimizations take care of it. Maybe I'm wrong and that is expensive.